src/ast/scopes.cc

```diff
@@ -376,6 +376,15 @@ void Scope::SetDefaults() {
   num_heap_slots_ = ContextHeaderLength();
 
   set_language_mode(LanguageMode::kSloppy);
+
+  // While replaying, tell v8 that inner scopes contain an eval().  This will
+  // keep scope analysis from optimizing away unreferenced variables in
+  // closures.
+  // (RUN-2604)
+  if (recordreplay::IsReplaying() &&
+      !recordreplay::FeatureEnabled("optimize-away")) {
+    inner_scope_calls_eval_ = true;
```

Comment:
this looks great! Are there any other downsides besides performance?

Reply:
not that I can _see_.  any direct reference to an outer binding would already cause it to be present, so semantically this feels safe.

I'll do some more testing of other scopes that v8 creates (default function arguments, e.g.) to make sure there's nothing visibly different.